luarules/gadgets/unit_seismic_dots_and_advanced_radar.lua

```diff
@@ -0,0 +1,296 @@
+			local unitPointX, unitPointY, unitPointZ = spGetUnitPosition(unitID) -- use basepoint as radar "paints" the ground
+			local dist2D = (basePointX-unitPointX)^2 + (basePointZ-unitPointZ)^2
+			local dist3D = (basePointX-unitPointX)^2 + (basePointY+radarEmitHeight-unitPointY-1)^2 + (basePointZ-unitPointZ)^2
+
+			-- check if the radar can see the unit without being obstructed by terrain
+			local raylength,_,_,_ = spTraceRayGroundBetweenPositions(basePointX, basePointY+radarEmitHeight, basePointZ, unitPointX, unitPointY + 1, unitPointZ, false)
+			if raylength == nil then -- returns nil if no ground collision detected
+				advancedRadarFreeLineOfFire = ((radarRadius+unitSpeeds[unitID])^2 >= dist2D)
+			end
+			-- +1 on unitPointX to prevent false ground collision positives at the terminal point.
+			-- +unitspeed on radarRadius to help catch units just walking into radar range, it seems unitspeed is accounted for when determining when to trigger UnitEnteredRadar
+
+			if advancedRadarFreeLineOfFire == true then 
+				wobblereduction = math.max(wobblereduction,tableValues.wobblereduction) -- use best wobblereduction of all advanced radars in range
+			end
+		end
+	end
+
+	if wobblereduction>0 then
+
+		if restoreUnitLOS[unitID] == nil then -- don't call customWobble if recently seismic pinged, customWobble will be called after unit LOS control is restored to engine
+			customWobble(unitID,advancedRadarUpdateRate,1-wobblereduction)
+		end
+
+		local framecycle = spGetGameFrame()%advancedRadarUpdateRate
+
+		inAdvancedRadar[allyTeam][framecycle][unitID] = true -- check this unit on next framecycle if it is still seen by an advanced radar
+		unitInAdvancedRadar[allyTeam][unitID] = framecycle -- lookup table for when a unit is scheduled to check for advanced radars again. Checked later to prevent a unit from being double scheduled
+		unitInBasicRadar[allyTeam][unitID] = nil -- effectively remove unit from the faster updating InBasicRadar list.
+
+	else
+		local framecycle = spGetGameFrame()%basicRadarUpdateRate
+		inBasicRadar[allyTeam][framecycle][unitID] = true -- check this unit on next framecycle if it is still seen by an advanced radar
+		unitInBasicRadar[allyTeam][unitID] = framecycle -- lookup table for when a unit is scheduled to check for advanced radars again. Checked later to prevent a unit from being double scheduled
+		unitInAdvancedRadar[allyTeam][unitID] = nil -- effectively remove unit from the slower updating InAdvancedRadar list.
+	end
+end
+
+function gadget:UnitEnteredRadar(unitID, unitTeam, allyTeam, unitDefID)
+
+	advancedRadarCheck(unitID,allyTeam) -- figure out if the unit entered a T1 or T2 radar range
+
+end
+
+function gadget:UnitLeftRadar(unitID, unitTeam, allyTeam, unitDefID)
+	-- effectively remove unit from the InRadar lists
+	unitInAdvancedRadar[allyTeam][unitID] = nil
+	unitInBasicRadar[allyTeam][unitID] = nil
+
+end
+
+function gadget:GameFrame(nn)
+
+	local advancedRadarUpdateFrame = (nn+1)%advancedRadarUpdateRate
```

Comment:
why the +1?

Reply:
Seems GameFrame is run at the very start of a sim frame, for all practical purposes. So I could remove the +1